x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/realm/TransportClearRealmCacheActionTests.java

```diff
@@ -0,0 +1,179 @@
+        verify(authenticationService).expire(user);
+    }
+
+    public void testSingleUserCacheCleanupForSingleRealm() {
+
+        final String user = "test";
+
+        // We want to clear user only from native realm cache.
+        final String[] realmsToClear = new String[] { nativeRealm.name() };
+        final String[] usersToClear = new String[] { user };
+        ClearRealmCacheRequest.Node clearCacheRequest = mockClearCacheRequest(realmsToClear, usersToClear);
+
+        ClearRealmCacheResponse.Node response = action.nodeOperation(clearCacheRequest, mock(Task.class));
+        assertThat(response, notNullValue());
+
+        // We expect that only native cache is cleared,
+        // including last successful cache in the authentication service.
+        verify(nativeRealm).expire(user);
+        verify(fileRealm, times(0)).expire(user);
+        verify(authenticationService).expire(user);
+    }
+
+    public void testAllUsersCacheCleanupForSingleRealm() {
+
+        // We want to clear all users from native realm cache.
+        final String[] realmsToClear = new String[] { nativeRealm.name() };
+        final String[] usersToClear = new String[] {};
+        ClearRealmCacheRequest.Node clearCacheRequest = mockClearCacheRequest(realmsToClear, usersToClear);
+
+        ClearRealmCacheResponse.Node response = action.nodeOperation(clearCacheRequest, mock(Task.class));
+        assertThat(response, notNullValue());
+
+        // We expect that whole native cache is cleared,
+        // including last successful cache in the authentication service.
+        verify(nativeRealm).expireAll();
+        verify(fileRealm, times(0)).expireAll();
+        verify(authenticationService).expireAll();
+    }
+
+    public void testAllUsersCacheCleanupForAllRealms() {
+
+        // We want to clear all users from all realms.
+        final String[] realmsToClear = new String[] {};
+        final String[] usersToClear = new String[] {};
+        ClearRealmCacheRequest.Node clearCacheRequest = mockClearCacheRequest(realmsToClear, usersToClear);
+
+        ClearRealmCacheResponse.Node response = action.nodeOperation(clearCacheRequest, mock(Task.class));
+        assertThat(response, notNullValue());
+
+        verify(fileRealm).expireAll();
+        verify(nativeRealm).expireAll();
+        verify(authenticationService).expireAll();
+    }
+
+    private TransportClearRealmCacheAction mockClearRealmCacheAction(Realms realms, AuthenticationService authenticationService) {
+        return new TransportClearRealmCacheAction(
+            mock(ThreadPool.class),
+            mockClusterService(),
+            mock(TransportService.class),
+            mock(ActionFilters.class),
+            realms,
+            authenticationService
+        );
+    }
+
+    private TestCachingRealm mockRealm(String name) {
+        TestCachingRealm realm = mock(TestCachingRealm.class);
+        when(realm.name()).thenReturn(name);
+        return realm;
+    }
```

Comment:
Unfortunately, we can't. Reason is because `Realms` holds `Realm` objects not `CachingRealm`. 
I thought of using `CachingUsernamePasswordRealm` instead, but this abstract class has made `expire` and `expireAll` methods final. Hence why I decided to introduce `TestCachingRealm`.

Reply:
All good. Thanks for the explanation.